### PR TITLE
route: add nexthop type

### PR DESCRIPTION
**Pull request by zstas**
The basic support for dumping nexthops and filling the cache.

### CODE REVIEW

File: lib/route/nh.c
Comment (thom311) [inline lib/route/nh.c:62]: there is only one caller of the function, and that one already checks that the refcount is not positive. This code can never be reached.

Also, a logging about a bug is not great. Let's do `_nl_assert(nhg->ce_refcnt == 0);`

(note that the caller actually only checks that the ce_refcnt is `<= 0`. But it really can also never be negative, that would be a bug too, and the assert will be more strict than the caller).

Same for the check of `if (!nhg)`. This is a static function (called from a few lines below). You can easily see, that the condition can never be true. An `_nl_assert(nhg)` would be more appropriate.


`_nl_assert()` is usually disabled in release builds (unless the user builds with `CFLAGS=-DNL_MORE_ASSERTS=1000`, which is uncommon).

File: lib/route/nh.c
Comment (thom311) [inline lib/route/nh.c:137]: I think a lot of the debug logging (all?) should be dropped. It doesn't seem useful when a library logs to stdout/stderr.

**Review by thom311 — CHANGES_REQUESTED** [review]
thank you for your patch. I am sorry for the long time it took me to reply. If you lost interest, I understand :( Anyway. Some comments...

File: libnl-cli-3.sym
Comment (zstas) [reply libnl-cli-3.sym:128]: done

File: libnl-route-3.sym
Comment (zstas) [reply libnl-route-3.sym:1302]: done

File: src/nl-nh-list.c
Comment (zstas) [reply src/nl-nh-list.c:30]: done

File: include/linux-private/linux/nexthop.h
Comment (zstas) [reply include/linux-private/linux/nexthop.h:5]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c]: done

**Comment by zstas** [comment]
H Thamas,

Thanks for the review. I tried to address all of them. Also I included another fix in the parser (in the previous patch only 2 group elements were allowed, I've increased it to 64). But unfortunately after rebasing github doesn't allow to compare the previous version with the current. But I've splitted the PR into 4 different commits (as you suggested), perhaps it makes the next review more easy.

I've been using the feature introduced in this patch for almost a year (for parsing fdb nexthops in linux), so I'd like to have this patch merged into upstream. So, I'm ready to improve this patch until you say it's good to go :)

File: include/netlink-private/types.h
Comment (thom311) [inline include/netlink-private/types.h:258]: this needs to go the the previous commit (2 of 4). Otherwise, that commit does not compile.

File: lib/route/neigh.c
Comment (thom311) [inline lib/route/neigh.c:524]: needs `"%u"` because `n->n_nhid` is `uint32_t`.

Technically, in plain C it might be that `uint32_t` is not the same as `unsigned int`, and thus `"%u"` would not be right. However, we rely on `sizeof(unsigned) >= sizeof(uint32_t)`, so `"%u"` is actually correct. To my knowledge, Linux kernel requires that `int` is at least 32bit(??).

File: include/netlink/route/neighbour.h
Comment (thom311) [reply include/netlink/route/neighbour.h:83]: The setter should get a `uint32_t` argument (note that libnl3 headers already include `<stdint.h>`, so we are fine to use that in public headers.

Also, the getter needs to return a full `uint32_t`.
```c
uint32_t rtnl_neigh_get_nhid(struct rtnl_neigh *);
```
might work, but then there is no room for an error code.

I think, this needs to be
```c
int rtnl_neigh_get_nhid(struct rtnl_neigh *neight, uint32_t *out_val);
```
where it return 0 on success or a negative `NLE_*` errror code.

File: lib/route/neigh.c
Comment (thom311) [inline lib/route/neigh.c:1091]: `return -NLE_NOATTR;`

Comment (thom311) [reply lib/route/neigh.c:1091]: can we do the error-paths first and return-early (not if/else)?
```c
    if (!(neigh->ce_mask & NEIGH_ATTR_NHID))
        return -NLE_NOATTR;
    return neigh->n_nhid;
```

File: include/netlink/route/neighbour.h
Comment (thom311) [inline include/netlink/route/neighbour.h:82]: does the patch lack a
```c
    if (tmpl->ce_mask & NEIGH_ATTR_NHID)
        NLA_PUT_U32(msg, NDA_NH_ID, tmpl->n_nhid);
```

?

otherwise, there is no point of having `rtnl_neigh_set_nhid()` in the public API.

File: Makefile.am
Comment (thom311) [inline Makefile.am:113]: I added a script to automatically format new source files with `clang-format`. There is also a github action that tests the format, but it was broken. If you would rebase to latest `main`, you would see that the github test `clang-format` fails due to that.

If you would, please reformat the files. As the tool would tell you:
```
Fix it by running `"./tools/clang-format.sh"` using clang-format version 16.0.6 (Fedora 16.0.6-2.fc38)
Alternatively, run `./tools/clang-format-container.sh` to use a podman container named "libnl-code-format-f38".
```

---

If that is cumbersome for you (e.g. because you don't use Fedora 38 with the right clang-format version and don't have podman working), you can also ignore that, then I will reformat your patch before merging. Just ignore it.

File: libnl-cli-3.sym
Comment (thom311) [inline libnl-cli-3.sym:130]: let's drop the empty line there.

File: libnl-route-3.sym
Comment (thom311) [inline libnl-route-3.sym:1302]: please sort the lines lexicographically.

File: include/netlink-private/types.h
Comment (thom311) [inline include/netlink-private/types.h:350]: some of those fields are unused (not yet implemented?). For example, `nh_protocol`.

It would seem clearer to just drop those. Otherwise, it prompts questions like whether `nh_clone()` needs to take care of those fields.

File: lib/route/nh.c
Comment (thom311) [inline lib/route/nh.c:92]: please add blank line before code.

File: lib/route/nh.c
Comment (thom311) [inline lib/route/nh.c:100]: please add blank line before code.

File: lib/route/nh.c
Comment (thom311) [inline lib/route/nh.c:189]: `return -NLE_NOMEM;`

**Review by thom311 — CHANGES_REQUESTED** [review]

File: include/netlink/route/neighbour.h
Comment (zstas) [reply include/netlink/route/neighbour.h]: done

File: lib/route/neigh.c
Comment (zstas) [reply lib/route/neigh.c]: done

File: lib/route/neigh.c
Comment (zstas) [reply lib/route/neigh.c]: done

File: include/netlink/route/neighbour.h
Comment (zstas) [reply include/netlink/route/neighbour.h]: I've added population the attribute.

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c:102]: done

File: lib/route/nh.c
Comment (zstas) [reply lib/route/nh.c:92]: done

File: include/netlink-private/types.h
Comment (zstas) [reply include/netlink-private/types.h:348]: done

File: libnl-route-3.sym
Comment (zstas) [reply libnl-route-3.sym:1301]: done

File: Makefile.am
Comment (zstas) [reply Makefile.am:113]: after a couple of tries it worked with my clang-format-14

File: include/netlink-private/types.h
Comment (zstas) [reply include/netlink-private/types.h:258]: done

**Comment by zstas** [comment]
I've addressed all the previous comments.
Also apologies for another last-minute change - since you mentioned nh_clone, I took another look and found that I wasn't copying nh_gateway and nh_group. Addressed that as well (in 3rd commit).

**Review by thom311 — CHANGES_REQUESTED** [review]
Maybe you could take the patches from https://paste.centos.org/view/3c278476 (well, review them, and take the parts that you agree with).

It also requires you to rebase the branch again to latest `main` (because they use some new API from `main` branch).

**Comment by zstas** [comment]
I intended to merge things from your patch today, but unfortunately the link is expired. Could you please share it one more time?

File: lib/route/neigh.c
Comment (zstas) [reply lib/route/neigh.c:737]: No, user-space must create a nexthop first (with a DST or a group, for vxlan also with flag `fdb`) and then create a neighbour entry with corresponding NH_ID. Dumping works the same way as it was created.

So, IMO it's better to skip that check and leave it to kernel.

**Comment by thom311** [comment]
I didn't realize fpaste expires so fast. Sorry about that. Here again: https://paste.centos.org/view/bba2d453

I also just realized, that a change I did will cause a merge conflict when you rebase. Sorry about that. It should however be easy to resolved... Thanks.

**Comment by zstas** [comment]
I rebased, applied all your changes and also removed nh_master attribute from `rtnl_nh` struct (as it wasn't used anywhere). Could you please take a look once again?

**Comment by thom311** [comment]
merged as bc8573d035eeea37d53d715a342111176f29fb19.

Thank you for your efforts!!!

**Pull request closed without merging**
